### PR TITLE
Core creds plugin alternative

### PR DESCRIPTION
**Pull request by jboeuf**
- The notification that the metadata is ready is now a fixed callback.
- This design is less flexible (especially for testing) but may be
  easier to integrate for wrapped languages.

### CODE REVIEW

File: include/grpc/grpc_security.h [hunk 1]
```diff
@@ -131,6 +131,34 @@ grpc_credentials *grpc_google_iam_credentials_create(
     const char *authorization_token, const char *authority_selector,
     void *reserved);
 
+/* Function to be called by the metadata credentials plugin implementation when
+   the metadata is ready. */
+void grpc_credentials_plugin_metadata_notify(
```
Comment (jtattermusch) [inline include/grpc/grpc_security.h]: naming seems inconsistent with grpc_metadata_credentials_plugin and grpc_metadata_credentials_create_from_plugin. Would any of these be better?
grpc_metadata_credentials_notify()
grpc_metadata_credentials_plugin_notify()  /* that's my favorite */
grpc_metadata_credentials_metadata_notify()

File: include/grpc/grpc_security.h [hunk 2]
```diff
@@ -131,6 +131,34 @@ grpc_credentials *grpc_google_iam_credentials_create(
...
+    void *core_context, const grpc_metadata *creds_md, size_t num_creds_md,
```
Comment (jtattermusch) [inline include/grpc/grpc_security.h]: By calling this, core becomes the owner of metadata keys and values and caller owns the grpc_metadata array?

Comment (jboeuf) [inline include/grpc/grpc_security.h]: No ownership is transferred here. Values are copied. Will update doc.

File: include/grpc/grpc_security.h (hunk 1 above)
Comment (jboeuf) [inline include/grpc/grpc_security.h]: The problem with `grpc_metadata_credentials_plugin_notify` is that this implies that this is a method of the `grpc_metadata_credentials_plugin` (you would expect a pointer to this type as the first param).
Howe about `grpc_metadata_credentials_notify_from_plugin`

Comment (jtattermusch) [inline include/grpc/grpc_security.h]: Sg. I mostly care about the prefixes being the same so that it's obvious that these things belong together.

**Comment by jtattermusch** [comment]
Is this PR in a good working shape? I'd like to try prototyping C# wrapping around the C core API to see how would the result feel.

**Comment by jboeuf** [comment]
I haven't tested anything so I cannot really guarantee that it works but I'm moderately confident that it will.

**Comment by murgatroid99** [comment]
For the record, I do not have a preference between this option and the other in terms of wrapping it in Node. It looks like it will just be the difference between passing around an opaque pointer and passing around two opaque pointers hidden by a single opaque pointer.

**Comment by jboeuf** [comment]
Abandoning the alternative as agreed with the team.

**Pull request closed without merging**